### DISCUSSION TITLE
Credential storage on MacOs/MacCatalyst

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#14773] [Q&A]
My app needs to store credentials of the logged in user. As per [this documentation](https://platform.uno/docs/articles/features/PasswordVault.html) it is not supported on Mac.

1. Is there a difference in this API regarding MacOS and MacCatalyst?
2. What is the recommended way of storing a credential on MacOs?

The doc is a bit confusing, too. `PasswordVault` is not supported, but `PasswordCredential` is. But I need `PasswordVault `for `PasswordCredential`? How does that come together?

**Comment by jeromelaban** [comment]
Thanks for the report. `PasswordVault` is supported on Catalyst, indeed the documentation does not accound for the addition of catalyst. Unless explicitly stated as Catalyst, macOS means AppKit, and the lact of Catalyst can be assumed as iOS.

**Reply by TopperDEL** [reply]
Hm, Do I need some permissions added or so? It seems to not store value here. Or is it only working together with `PasswordCredential`? Need to check tomorrow.

**Reply by TopperDEL** [reply]
In my case it seems to write (or read) empty credentials. It seems to not work an Mac Catalyst.

**Reply by TopperDEL** [reply]
This is my LocalSecretStorageService:

```
public class LocalSecretsStore : ILocalSecretsStore
 {
     private const string RESOURCE = "MySecrets";

     public void ClearSecret(string secretKey)
     {
         try
         {
             var vault = new Windows.Security.Credentials.PasswordVault();
             var credential = vault.Retrieve(RESOURCE, secretKey);
             vault.Remove(credential);
         }
         catch
         {
             //Ignore
         }
     }

     public string GetSecret(string secretKey)
     {
         try
         {
             var vault = new Windows.Security.Credentials.PasswordVault();
             var credential = vault.Retrieve(RESOURCE, secretKey);
             credential.RetrievePassword();
             return credential.Password;
         }
         catch
         {
             return string.Empty;
         }
     }

     public bool SetSecret(string secretKey, string secretValue)
     {
         try
         {
             var vault = new Windows.Security.Credentials.PasswordVault();
             vault.Add(new Windows.Security.Credentials.PasswordCredential(RESOURCE, secretKey, secretValue));
             return true;
         }
         catch
         {
             return false;
         }
     }
 }
```

**Reply by jeromelaban** [reply]
Can you detail specifically what does not work? Any exception, log that you can provide? Let's open an issue with those details please.

**Reply by TopperDEL** [reply]
I currently have problems debugging on Mac. But from the error I get the saved secret returns an empty string. I try to parse that string to JSON and get something like "no tokens in JSON".
Is there maybe a length restriction?

**Reply by TopperDEL** [reply]
Ok, I get the error on save that my app is not allowed to access the KeyChain and that I have to add it to the Entitlements.plist. This should be added to the documentation, I might create a PR for that.

BUT: it is still not working if I add an entry. My Entitlements.plist for MacCatalyst looks like this:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>keychain-access-groups</key>
  <array>
    <string>$(AppIdentifierPrefix)$(CFBundleIdentifier)</string>
  </array>

</dict>
</plist>
```

I also tried a fixed value with the app bundle name ("com.myapp"). Neither is working, I always get the error.

**Reply by jeromelaban** [reply]
Can you provide the actual error you're getting?

**Reply by TopperDEL** [reply]
The exact error message is: `Your application is not allowed to use the keychain. Make sure that you have setup the KeyChain in the Entitlements.plist of your application.`

I also tried adding a `<CodesignEntitlements>MacCatalyst\Entitlements.plist</CodesignEntitlements>`-entry to the csproj - but then the app does not startup anymore with a "The application XX could not be opened"-Message.

**Reply by jeromelaban** [reply]
Have you added the keychain option in your provisioning profile as well?

**Reply by TopperDEL** [reply]
Does that need to be done in the apple developer account?

**Reply by TopperDEL** [reply]
I see I did not have the app-name registered and have no specific provisioning profile for the app. Maybe that's the reason then. I will investigate this a bit more.

**Reply by TopperDEL** [reply]
Hm, I still don't get it. I registered my app name, I've created a provisioning profile and checked that Xcode and vs4mac have access to my apple developer account. It does not make a difference.

**Reply by jeromelaban** [reply]
@spouliot would you have an idea of what could happen here?

**Reply by spouliot** [reply]
Check the entitlements inside the `.app`. Sometimes settings can be invalid/incomplete and the build won't show an error, but it won't work as expected at runtime.

Also open `Console.app` to see what the OS is logging. There's often a few good hints about why things are not working.

**Reply by TopperDEL** [reply]
Hm. I cannot find the Entitlements inside the `.app` - where should they be? But this strengthens my opinion, that the Entitlements.plist is not added.
[This](https://learn.microsoft.com/en-us/dotnet/maui/mac-catalyst/entitlements?view=net-maui-8.0) link shows how to add the Entitlements.plist and the keychain capability to the project. I tried this but there is still no change - I cannot find any Entitlements.plist int he app package and I still do not have the permission.

I also tried the `Console.app` - did not know that before. But I also cannot find any "access restriction" error there. Do you know what I would have to check?

I also tried to extract the Entitlements from the app with commands found via Google, but none of them works as expected.

Do you have an app with that capability, that I may check against my setup?

**Reply by TopperDEL** [reply]
Oh wait - I added it only to the debug version but was checking the release version. Now I can extract the Entitlement using `codesign -d --entitlements :- MyApp.app`. And I can see my `keychain-access-groups`-entry.
But then once again the app does not startup. "The application 'MyApp' could not be opened." (German text is "Das Programm 'MyApp' kann nicht geöffnet werden."

**Reply by spouliot** [reply]
> Now I can extract the Entitlement using codesign -d --entitlements :- MyApp.app. And I can see my keychain-access-groups-entry.

Great!

> "The application 'MyApp' could not be opened."

This is a window popup, right ?

Console.app should have logs about this. Try to filter the logs with your application name (or bundle id).

**Reply by TopperDEL** [reply]
Yes, this is a window popup.

I still do not have a clue what's happening. I do not get a helpful message except "it is not allowed".

**Reply by spouliot** [reply]
I look into a [similar issue](https://github.com/unoplatform/uno.extensions/issues/2126) today but it never showed me a window popup, the app just hang on start up (so it likely no 100% identical). However there were _hints_ to be seen in the logs inside Console.app.

**Reply by TopperDEL** [reply]
Hm, I did not get it solved. I switched to another unsave implementation using a local file just for getting any further but left the Mac-App rest until some other issues are fixed. So it is not immediate important, currently. But I did not find a solution, yet.
